pkg/beaconfqdn/dissector.go

```diff
@@ -0,0 +1,174 @@
+}
+
+//close waits for the collector to finish
+func (d *dissector) close() {
+	close(d.dissectChannel)
+	d.dissectWg.Wait()
+	// d.closedCallback()
+}
+
+//start kicks off a new analysis thread
+func (d *dissector) start() {
+	d.dissectWg.Add(1)
+	go func() {
+		ssn := d.db.Session.Copy()
+		defer ssn.Close()
+
+		for datum := range d.dissectChannel {
+
+			// set up src match key (src ip + network uuid)
+			srcMatchKey := datum.Src.SrcBSONKey()
+
+			var dstList []interface{}
+			// create dst match query section
+			for _, dst := range datum.ResolvedIPs {
+				dstList = append(dstList, dst.DstBSONKey())
+			}
+
+			// This will work for both updating and inserting completely new Beacons
+			// for every new hostnames record we have, we will check every entry in the
+			// uconn table where the source IP from the hostnames record connected to one
+			// of the associated IPs for  FQDN. This
+			// will always return a result because even with a brand new database, we already
+			// created the uconns table. It will only continue and analyze if the connection
+			// meets the required specs, again working for both an update and a new src-fqdn
+			// pair. We would have to perform this check regardless if we want the rolling
+			// update option to remain, and this gets us the vetting for both situations, and
+			// Only works on the current entries - not a re-aggregation on the whole collection,
+			// and individual lookups like this are really fast. This also ensures a unique
+			// set of timestamps for analysis.
+			uconnFindQuery := []bson.M{
+				bson.M{"$match": srcMatchKey},
+				bson.M{"$match": bson.M{"$or": dstList}},
+				bson.M{"$project": bson.M{
+					"src": 1,
+					"ts": bson.M{
+						"$reduce": bson.M{
+							"input":        "$dat.ts",
+							"initialValue": []interface{}{},
+							"in":           bson.M{"$concatArrays": []interface{}{"$$value", "$$this"}},
+						},
+					},
+					"bytes": bson.M{
+						"$reduce": bson.M{
+							"input":        "$dat.bytes",
+							"initialValue": []interface{}{},
+							"in":           bson.M{"$concatArrays": []interface{}{"$$value", "$$this"}},
+						},
+					},
+					"count":  bson.M{"$sum": "$dat.count"},
+					"tbytes": bson.M{"$sum": "$dat.tbytes"},
+					"icerts": bson.M{"$anyElementTrue": []interface{}{"$dat.icerts"}},
+				}},
+				bson.M{"$group": bson.M{
+					"_id":    "$src",
+					"ts":     bson.M{"$push": "$ts"},
+					"bytes":  bson.M{"$push": "$bytes"},
+					"count":  bson.M{"$sum": "$count"},
+					"tbytes": bson.M{"$sum": "$tbytes"},
+					"icerts": bson.M{"$push": "$icerts"},
+				}},
+				bson.M{"$unwind": "$ts"},
+				bson.M{"$unwind": "$ts"},
+				bson.M{"$group": bson.M{
+					"_id":    "$_id",
+					"ts":     bson.M{"$addToSet": "$ts"},
+					"bytes":  bson.M{"$first": "$bytes"},
+					"count":  bson.M{"$first": "$count"},
+					"tbytes": bson.M{"$first": "$tbytes"},
+					"icerts": bson.M{"$first": "$icerts"},
+				}},
+				bson.M{"$unwind": "$bytes"},
```

Comment:
I think you're right! But lets make a separate issue for that, since we would need to test and update the other version of beaconing as well. Right now we want this one to match the other one.